### PR TITLE
namespace controller: use shared informer

### PR DESCRIPTION
**Pull request by howardjohn**
This updates the namespace controller to use the new shared informers. This is a perf optimization to minimize the number of watches we open to the api server.

Part of https://github.com/istio/istio/issues/24640

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go [hunk 1]
```diff
@@ -56,59 +50,42 @@ type NamespaceController struct {
...
 // NewNamespaceController returns a pointer to a newly constructed NamespaceController instance.
-func NewNamespaceController(data func() map[string]string, options Options, kubeClient kubernetes.Interface) *NamespaceController {
+func NewNamespaceController(data func() map[string]string, kubeClient kube.Client) *NamespaceController {
 	c := &NamespaceController{
 		getData: data,
 		client:  kubeClient.CoreV1(),
 		queue:   queue.NewQueue(time.Second),
 	}
 
-	watchedNamespaceList := strings.Split(options.WatchedNamespaces, ",")
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go:75]: Is multi-namespaces not supported though this is not perfect now?

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go:75]: There is work to add this back correctly, see https://github.com/istio/istio/issues/24640#issuecomment-654309550

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go:75]: which will be done before 1.7

File: pilot/pkg/bootstrap/server.go [hunk 2]
```diff
@@ -999,11 +999,17 @@ func (s *Server) initNamespaceController(args *PilotArgs) {
...
 					log.Infof("Starting namespace controller")
-					nc := kubecontroller.NewNamespaceController(s.fetchCARoot, args.RegistryOptions.KubeOptions, s.kubeClient)
-					nc.Run(stop)
+					nc := kubecontroller.NewNamespaceController(s.fetchCARoot, s.kubeClient)
+					// Start informers again. This fixes the case where informers for namespace do not start,
+					// as we create them only after acquiring the leader lock
+					// Note: stop here should be the overall pilot stop, NOT the leader election stop. We are
+					// basically lazy loading the informer, if we stop it when we lose the lock we will never
+					// recreate it again.
+					s.kubeClient.RunAndWait(stop)
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/server.go:1011]: can me move the namespace informer into `nc.Run`. It is more clear that we donot need to care about other informers

File: pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go [hunk 3]
```diff
@@ -56,59 +50,42 @@ type NamespaceController struct {
...
-
-	mlw := listwatch.MultiNamespaceListerWatcher(watchedNamespaceList, func(namespace string) cache.ListerWatcher {
-		return &cache.ListWatch{
-			ListFunc: func(opts metav1.ListOptions) (runtime.Object, error) {
-				opts.LabelSelector = fields.SelectorFromSet(configMapLabel).String()
-				return kubeClient.CoreV1().ConfigMaps(namespace).List(context.TODO(), opts)
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go:81]: just realized this watch only the specific configmap, with this pr we watch all. The difference lays on the memory store.

**Pull request merged**